### PR TITLE
Limit loopStrider to children of array-shadow symbol nodes

### PR DESCRIPTION
**Pull request by rmnattas**
LoopStrider currently performs on non-arrays and that's was not the intended design for it. This sets non-array nodes as pinning-array and the access as an array access.
This didn't cause an issue previously until OffHeap was introduced. Given that OffHeap array accesses requires loading the `dataAddrPtr`, after inserting the load to a non-array access, it ends up loading/storing to a different memory location.

The fix limits loopStrider to the first-level children of the array-shadow symbol nodes.

An extra check against unsafe-shadow symbols is added as unsafe-shadow symbols also have array-shadow symbol flag set.

Before the fix:
```
n25090n   astore  <temp slot 68>[#3108  Auto] [flags 0x7 0x0 ]                                [0x788beb709240] bci=[102,3,251] rc=0 vc=4560 vn=19 li=287 udi=3 nc=1
n25089n     aload  CurrentThread[#343  MethodMeta +160] [flags 0x207 0x400 ]                  [0x788beb7091f0] bci=[102,3,251] rc=1 vc=4560 vn=19 li=- udi=- nc=0
...
n6799n      iloadi  <unsafe shadow sym>[#1426  Shadow] [flags 0x80000603 0x100 ]              [0x788c01563cf0] bci=[4,9,232] rc=1 vc=4560 vn=54 li=- udi=- nc=1
n6812n        aladd (X>=0 internalPtr )                                                       [0x788c01564100] bci=[4,3,232] rc=1 vc=4560 vn=53 li=- udi=- nc=2 flg=0x8100
n6793n          aload  <temp slot 68>[#3108  Auto] [flags 0x7 0x0 ] (X!=0 createdByPRE )      [0x788c01563b10] bci=[4,3,232] rc=1 vc=4560 vn=19 li=- udi=642 nc=0 flg=0x40004
n6797n          lconst 108 (highWordZero X!=0 X>=0 )                                          [0x788c01563c50] bci=[4,6,232] rc=1 vc=4560 vn=52 li=- udi=- nc=0 flg=0x4104

loopStrider

n25090n   astore  <pinning array temp slot 68>[#3108  Auto] [flags 0x10000007 0x0 ]           [0x788beb709240] bci=[102,3,251] rc=0 vc=4560 vn=- li=287 udi=3 nc=1
n25089n     aload  CurrentThread[#343  MethodMeta +160] [flags 0x207 0x400 ]                  [0x788beb7091f0] bci=[102,3,251] rc=1 vc=4560 vn=- li=- udi=- nc=0
...
n26905n   astore  <internal pointer temp slot 93>[#3133  Auto] [flags 0x40007 0x0 ]           [0x788beaefc990] bci=[-1,5,170] rc=0 vc=0 vn=- li=-1 udi=- nc=1
n26904n     aladd (internalPtr )                                                              [0x788beaefc940] bci=[-1,5,170] rc=1 vc=0 vn=- li=- udi=- nc=2 flg=0x8000
n26903n       aloadi  <contiguousArrayDataAddrField>[#374  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [0x788beaefc8f0] bci=[-1,5,170] rc=1 vc=0 vn=- li=- udi=- nc=1 flg=0x8000
n26902n         aload  <pinning array temp slot 68>[#3108  Auto] [flags 0x10000007 0x0 ]      [0x788beaefc8a0] bci=[-1,5,170] rc=1 vc=0 vn=- li=- udi=- nc=0
n26901n       lconst 108 (highWordZero X!=0 X>=0 )                                            [0x788beaefc850] bci=[-1,5,170] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
```

Closes: https://github.com/eclipse-openj9/openj9/issues/21620

### CODE REVIEW

**Comment by vijaysun-omr** [comment]
Is this only an issue in the reassociation code OR also in other places in the strider that actually do striding on induction variables ?

**Comment by rmnattas** [comment]
I'll look into `morphExpressionLinearInInductionVariable` and test it with something like `UNSAFE.get(obj,induction-offset)`

**Comment by rmnattas** [comment]
I added the array-shadow check to `examineTreeForInductionVariableUse` in a WIP way to test and tests are passing so-far (mid-way).

Issue with changing its definition to add the `arrayShadowParent` argument is that it's an override definition.

The `examineTreeForInductionVariableUse` is a virtual declaration under `TR_LoopTransformer` which has `TR_LoopStrider` and `TR_LoopCanonicalizer` subclasses, each with their own implementation of `examineTreeForInductionVariableUse`. Hence adding the argument to the only the loopStrider definition is not allowed. 
Reason is that both loopStrider and loopCanonicalizer call the same `TR_LoopTransformer::replaceAllInductionVariableComputations` which ends up calling different implemenations of `examineTreeForInductionVariableUse`.

Currently I'm using a default value for the argument and adding it to the loopCanonicalizer version but not using it as a way to make it work. But looking for a permanent way to solve it.

**Comment by vijaysun-omr** [comment]
You could use an instance field to pass the parent across invocations ? You can save and restore the field on all return paths to simulate what might have happened with argument passing. Thoughts ?

**Comment by rmnattas** [comment]
Some kind of a shared variable would work for both `reassociateAndHoistComputations` and `examineTreeForInductionVariableUse` too. I don't believe there's a case of `array-shadow` having more than one child (where the flag might need to be set for both to imitate recursion), and no execution overlap between the 2 methods using the variable.

**Comment by rmnattas** [comment]
PR ready for review

File: compiler/optimizer/InductionVariable.cpp [hunk 1]
```diff
@@ -2188,6 +2189,13 @@ bool TR_LoopStrider::examineTreeForInductionVariableUse(TR::Block *loopInvariant
          }
       }
 
+   // Only first-level childerent of array-shadow symbol node have the _arrayShadowParent argument set
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:2192]: typo, "childrent"

File: compiler/optimizer/InductionVariable.cpp [hunk 2]
```diff
@@ -2188,6 +2189,13 @@ bool TR_LoopStrider::examineTreeForInductionVariableUse(TR::Block *loopInvariant
...
+   if (_arrayShadowParent)
+      _arrayShadowParent = false;
+   TR::Node* checkNode = originalNode ? originalNode : node;
+   if (checkNode->getOpCode().hasSymbolReference() && checkNode->getSymbolReference()->getSymbol()->isArrayletShadowSymbol())
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:2196]: This should be a test on "array shadow symbol" rather than "arraylet shadow symbol" ? The latter won't even be seen in gencon and likely not even in off heap balanced GC, I think

File: compiler/optimizer/InductionVariable.cpp [hunk 3]
```diff
@@ -3739,6 +3747,13 @@ bool TR_LoopStrider::reassociateAndHoistComputations(TR::Block *loopInvariantBlo
       }
 #endif /* J9_PROJECT_SPECIFIC */
 
+   // Only first-level childerent of array-shadow symbol node have the _arrayShadowParent argument set
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:3750]: Same typo as earlier comment

File: compiler/optimizer/InductionVariable.cpp [hunk 4]
```diff
@@ -3739,6 +3747,13 @@ bool TR_LoopStrider::reassociateAndHoistComputations(TR::Block *loopInvariantBlo
...
+   if (_arrayShadowParent)
+      _arrayShadowParent = false;
+   TR::Node* checkNode = originalNode ? originalNode : node;
+   if (checkNode->getOpCode().hasSymbolReference() && checkNode->getSymbolReference()->getSymbol()->isArrayletShadowSymbol())
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:3754]: Same concern at the shadow being checked here as earlier comment

**Comment by vijaysun-omr** [comment]
Jenkins build all

**Comment by vijaysun-omr** [comment]
Looks fine to me. I want to confirm with @rmnattas that unlike the prior version of this fix, this version does result in striding in other places but obviously I would expect it avoid striding of non-array accesses. i.e. from looking at some log or trace output where we used to do striding on array accesses, we still are doing it.

**Comment by vijaysun-omr** [comment]
@zl-wang if you wanted to do a quick review too before I merge this after Abdul replies.

**Review by zl-wang — APPROVED** [review]
LGTM

**Comment by rmnattas** [comment]
Local testing shows loopStrider behaving as expected after the extra guard against unsafe shadow symbols.

**Comment by vijaysun-omr** [comment]
Jenkins build all

**Comment by rmnattas** [comment]
I looked at few typical array loop-striding and induction-variable expressions cases and the edge case bug issue.
Running jenkins tests too

**Comment by vijaysun-omr** [comment]
Okay, I will wait for your all clear from the additional testing you are running. I will merge once I have that, and once checks have passed on this PR.

**Comment by rmnattas** [comment]
Jenkins tests look good, ready for merge
